src/bptl/credentials/models.py

```diff
@@ -0,0 +1,143 @@
+import logging
+from typing import Dict
+
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+
+from zgw_consumers.constants import AuthTypes
+
+logger = logging.getLogger(__name__)
+
+
+class App(models.Model):
+    label = models.CharField(
+        _("label"), max_length=100, help_text=_("Human readable app identifier.")
+    )
+    app_id = models.CharField(
+        _("app ID"),
+        max_length=1000,
+        unique=True,
+        help_text=_(
+            "(Globally) unique application identifier. Typically the "
+            "URL of the application in the Autorisaties API."
+        ),
+    )
+
+    class Meta:
+        verbose_name = _("application")
+        verbose_name_plural = _("applications")
+
+    def __str__(self):
+        return self.label
+
+
+class AppServiceCredentials(models.Model):
+    """
+    Connect an App to a Service with app-specific credentials.
+    """
+
+    # relational data - connect app to service and vice versa
+    app = models.ForeignKey(
+        "App",
+        on_delete=models.CASCADE,
+        verbose_name=_("application"),
+    )
+    service = models.ForeignKey(
+        "zgw_consumers.Service",
+        on_delete=models.CASCADE,
+        verbose_name=_("service"),
+        help_text=_("External service to configure credentials for."),
+    )
+
+    # store the required credentials
+    # TODO: use encryption extensions for these fields
+
+    # ZGW client ID + secret
+    client_id = models.CharField(max_length=255, blank=True)
+    secret = models.CharField(max_length=255, blank=True)
+
+    # API key of sorts
+    header_key = models.CharField(_("header key"), max_length=100, blank=True)
+    header_value = models.CharField(_("header value"), max_length=255, blank=True)
```

Comment:
The same fields were picked on purpose to keep the similarity.

This is a multi-layered approach. In the case of Open Zaak clients, each has different credentials. However, if we have credentials for BAG or BRP API for example, it's much more likely that we only have a single API Key for the entire organization, and then you don't want to repeat that configuration.

So, what happens is:

* determine services
* check if there are credentials specific to the app (pointed at by `bptlAppId`), use them if they exist
* if there aren't, rely on the Service credentials, which are global to BPTL

You have a multi-staged approach to credentials this way with all the flexibility and the ease of not having to define the same credentials over and over again.

> Perhaps zgw_consumers.Service can be changed: for example, it can be split to base and proxy models with unique constraint only on proxy model (and in this case, it can be backward compatible).

I don't think that adding the extra complexity to zgw_consumers really solves anything - it's just not relevant for other applications using that library. BPTL has a very niche case of managing credentials, and imo the solution for that belongs in BPTL. By default storing the credentials in a separate model in zgw-consumers incurs an INNER_JOIN performance penalty, which is really not needed.

> This case with several credentials for one API can appear somewhere in the future again, I think we need to include its support to zgw_consumers

Let's reconsider if it does appear again in the future.

Reply:
Ok, I agree that the multi-layered approach fits BPTL case well.